Enable hostpolicy on VM mode for -enableKubeArmorHostPolicy flag

seswarrajan opened the pull request:
Signed-off-by: Eswar Rajan Subramanian <eswar@accuknox.com>

seswarrajan committed:
Enable hostpolicy on VM mode for -enableKubeArmorHostPolicy flag

Signed-off-by: Eswar Rajan Subramanian <eswar@accuknox.com>

seswarrajan committed:
Enable hostpolicy on VM mode

Signed-off-by: Eswar Rajan Subramanian <eswar@accuknox.com>

nyrahul commented on KubeArmor/common/common.go:
```diff
@@ -322,7 +322,7 @@ func GetExternalIPAddr() string {
-	if _, err := os.Stat(filepath.Clean(k8sConfig)); err == nil {
+	if _, err := os.Stat(filepath.Clean(k8sConfig)); k8sConfig != "" && err == nil {
 		return true
 	}
```
No point in doing a `os.Stat` and then checking if `k8sConfig != ""`!
```suggestion
        if k8sConfig != "" {
	    if _, err := os.Stat(filepath.Clean(k8sConfig)); err == nil {
	   	return true
	    }
	}
```

nyrahul commented on KubeArmor/core/kubeArmor.go:
```diff
@@ -321,7 +321,8 @@ func KubeArmor() {
 
-	if cfg.GlobalCfg.KVMAgent {
+	// Enable KubeArmorHostPolicy for both VM and KVMAgent
+	if !kl.IsK8sEnv() && (cfg.GlobalCfg.KVMAgent || cfg.GlobalCfg.HostPolicy) {
```
What happens in cases where user has `~/.kube/config` file, but still wants to use kubearmor in pure VM/systemd mode?
Just the presence of kubeconfig file will cause kubearmor to think that it is using k8s mode of operation!

seswarrajan replied:
Yes. Possible. In that case it will deny/ignore to run as systemd mode.

nam-jaehyun replied:
How can we know if a user wants to run KubeArmor as a systemd mode in such as case?
@seswarrajan What happens if a user uses -enableKubeArmorVm without kvm-service?

seswarrajan replied:
@nam-jaehyun User can enable KubeArmor systemd mode only by -enableKubeArmorHostPolicy flag. It is not possible to run kubeArmor in VM with -enableKubeArmorVm flag as this flag tries to connect to kvm-service and if the connection fails, KubeArmor will terminate.

seswarrajan replied:
@nyrahul In that case we need to ignore the IsK8sEnv() validation check in the code. But internally in the code while applying for new policies IsK8sEnv() check is still run.

nyrahul replied:
This has to be fixed holistically ... currently the code just allows to bypass the first hurdle. If the IsK8sEnv() check is present elsewhere, then VM/systemd mode cannot be used.

Can you check if kubearmor works in systemd mode (with policy enforcement) when there is ~/.kube/config in the path?

seswarrajan committed:
Addressed review comments

Signed-off-by: Eswar Rajan Subramanian <eswar@accuknox.com>

nam-jaehyun reviewed (approved):
I'm okay with this PR.

seswarrajan committed:
Merge branch 'kubearmor:main' into main

seswarrajan committed:
Merge branch 'kubearmor:main' into main

seswarrajan committed:
Added new flag non-k8s
This flag is by deafult false, needs to be specified as -non-k8s when
required to skip LocalK8s check

Signed-off-by: Eswar Rajan Subramanian <eswar@accuknox.com>

nyrahul commented on KubeArmor/common/common.go:
```diff
@@ -23,6 +23,8 @@ import (
 // == Common == //
 // ============ //
 
+var IsNonK8sEnv bool
```
Please fix the lint warnings.

seswarrajan committed:
Fixed lint warnings

Signed-off-by: Eswar Rajan Subramanian <eswar@accuknox.com>

seswarrajan committed:
Modified flag non-k8s to k8s

Signed-off-by: Eswar Rajan Subramanian <eswar@accuknox.com>

seswarrajan committed:
Modified Kubearmor to use K8S flag for VM/SYSTEMD modes

Signed-off-by: Eswar Rajan Subramanian <eswar@accuknox.com>

seswarrajan committed:
Enabled k8s flag for kubearmorVM/Host-only

Signed-off-by: Eswar Rajan Subramanian <eswar@accuknox.com>

nyrahul commented on KubeArmor/Makefile:
```diff
@@ -39,7 +39,7 @@ run: build
 	cd $(CRDDIR); kubectl apply -f KubeArmorHostPolicy.yaml
 	cd $(CURDIR); sudo rm -f /tmp/kubearmor.log
-	cd $(CURDIR); sudo -E ./kubearmor -logPath=/tmp/kubearmor.log -enableKubeArmorPolicy=false -enableKubeArmorHostPolicy=true
+	cd $(CURDIR); sudo -E ./kubearmor -logPath=/tmp/kubearmor.log -enableKubeArmorPolicy=false -enableKubeArmorHostPolicy=true -k8s=false
```
Cannot set k8s to false in the Makefile ... This will impact regular execution.

seswarrajan replied:
But this is got Kubearmor run-host-only. When run host-only is invoked via make and also .kube/config is present, then host mode will not work.

nyrahul commented on KubeArmor/config/config.go:
```diff
@@ -144,9 +150,11 @@ func LoadConfig() error {
 		GlobalCfg.HostPolicy = true
 	}
+	GlobalCfg.K8sEnv = viper.GetBool(ConfigK8sEnv)
+	kl.IsK8SEnv = viper.GetBool(ConfigK8sEnv)
```
Please store the IsK8sEnv in GlobalCfg and then use GlobalCfg in kl module.

seswarrajan replied:
It is not possible to include config in common as it causes circular dependency. Hence this global variable.

nyrahul commented on KubeArmor/config/config.go:
```diff
@@ -28,6 +28,7 @@ type KubearmorConfig struct {
 	Policy     bool // Enable/Disable policy enforcement
 	HostPolicy bool // Enable/Disable host policy enforcement
 	KVMAgent   bool // Enable/Disable KVM Agent
+	K8sEnv     bool // Is k8s env ?
```
This variable is not used currently. Please set this variable.

nyrahul commented on KubeArmor/common/common.go:
```diff
@@ -23,6 +23,9 @@ import (
 
+// IsK8SEnv used for specifying k8s environment
+var IsK8SEnv bool
+
```
```suggestion
```
Remove this var and use GlobalCfg.K8sEnv bool in the check below.

seswarrajan replied:
It is not possible to include config in common as it causes circular dependency. Hence this global variable.

nyrahul commented on KubeArmor/common/common.go:
```diff
@@ -329,9 +332,15 @@ func GetExternalIPAddr() string {
 
 // IsK8sLocal Function
 func IsK8sLocal() bool {
+	if !IsK8SEnv {
```
```suggestion
	if !cfg.K8sEnv {
```

seswarrajan replied:
It is not possible to include config in common as it causes circular dependency. Hence this global variable.

seswarrajan committed:
included k8s flag for specifying environment

Signed-off-by: Eswar Rajan Subramanian <eswar@accuknox.com>
